js/src/main/scala-2.10/scalajson.ast/JValue.scala

```diff
@@ -143,9 +142,85 @@ final class JNumber private[ast] (val value: String) extends JValue {
 
   def copy(value: String): JNumber =
     value match {
-      case jNumberRegex(_ *) => new JNumber(value)
+      case jNumberRegex(_ *) => new JNumber(value)(0)
       case _ => throw new NumberFormatException(value)
     }
+
+  def toInt: Option[Long] = {
+    if ((constructedFlag & NumberFlags.int) == NumberFlags.int)
+      Some(value.toInt)
+    else {
+      try {
+        val asInt = value.toInt
+        if (BigInt(value) == BigInt(asInt))
+          Some(asInt)
+        else
+          None
+      } catch {
+        case _: NumberFormatException => None
+      }
+    }
+  }
+
+  def toLong: Option[Long] = {
+    if ((constructedFlag & NumberFlags.long) == NumberFlags.long)
+      Some(value.toLong)
+    else {
+      try {
+        val asLong = value.toLong
+        if (BigInt(value) == BigInt(asLong))
+          Some(asLong)
+        else
+          None
+      } catch {
+        case _: NumberFormatException => None
+      }
+    }
+  }
+
+  def toBigInt: Option[BigInt] = {
+    if ((constructedFlag & NumberFlags.bigInt) == NumberFlags.bigInt)
+      Some(BigInt(value))
+    else {
+      try {
+        Some(BigInt(value))
+      } catch {
+        case _: NumberFormatException => None
+      }
+    }
+  }
+
+  def toBigDecimal: Option[BigDecimal] = {
+    try {
+      Some(BigDecimal(value))
+    } catch {
+      case _: NumberFormatException => None
+    }
+  }
+
+  def toFloat: Option[Float] = {
+    if ((constructedFlag & NumberFlags.float) == NumberFlags.float)
+      Some(value.toFloat)
+    else {
+      val asFloat = value.toFloat
+      if (BigDecimal(value) == BigDecimal(asFloat.toDouble))
```

Comment:
> Getting rid of `Float` for correctness is no more sensible than getting rid of `Double` for correctness.

Both `def toFloat: Option[Float]` and `def toDouble: Option[Double]` that adds shape checking (`BigDecimal(value) == BigDecimal(asFloat.toDouble)`) are ridiculous. Because IEEE floats internally represent decimals as a binary fraction, it can't express `0.1`.

If we are letting them survive, trusting that the user knows what they are doing, I think the Scala idiomatic way is to behave like StringOps and `def toFloat: Float` and `def toDouble: Double` return whatever `value.toFloat` / `value.toDouble` returns.